src/main/scala/workflow/Pipeline.scala

```diff
@@ -0,0 +1,242 @@
+      case (transformer: TransformerNode[_], id) => s"$id [label=${'"' + transformer.label + '"'}]"
+      case (estimator: EstimatorNode, id) => s"$id [label=${'"' + estimator.label + '"'} shape=${"\"box\""}]"
+    } :+ s"${nodes.size} [label=${"\"Out\""} shape=${"\"Msquare\""}]"
+
+    val dataEdges: Seq[String] = dataDeps.zipWithIndex.flatMap {
+      case (deps, id) => deps.map(x => s"$x -> $id")
+    } :+ s"$sink -> ${nodes.size}"
+
+    val fitEdges: Seq[String] = fitDeps.zipWithIndex.flatMap {
+      case (deps, id) => deps.map(x => s"$x -> $id [dir=${"\"none\""} style=${"\"dashed\""}]")
+    }
+
+    val ranks = fitDeps.zipWithIndex.flatMap {
+      case (deps, id) => deps.map(x => s"{rank=same; $x $id}")
+    }
+
+    val lines = nodeLabels ++ dataEdges ++ fitEdges ++ ranks
+    lines.mkString("digraph pipeline {\n  rankdir=LR;\n  ", "\n  ", "\n}")
+  }
+
+  final private[workflow] def planEquals(pipeline: Pipeline[A, B]): Boolean = {
+    this.eq(pipeline) || (
+        (nodes == pipeline.nodes) &&
+        (dataDeps == pipeline.dataDeps) &&
+        (fitDeps == pipeline.fitDeps) &&
+        (sink == pipeline.sink))
+  }
+}
+
+object Pipeline {
+  val SOURCE: Int = -1
+
+  /**
+   * Constructs the Identity pipeline
+   * @tparam T The type of input to take
+   */
+  def apply[T](): Pipeline[T, T] = new ConcretePipeline(Seq(), Seq(), Seq(), SOURCE)
+  private[workflow] def apply[A, B](nodes: Seq[Node], dataDeps: Seq[Seq[Int]], fitDeps: Seq[Seq[Int]], sink: Int): Pipeline[A, B] = new ConcretePipeline(nodes, dataDeps, fitDeps, sink)
+
+  /**
+   * Produces a pipeline that when given an input,
+   * combines the outputs of all its branches when executed on that input into a single Seq (in order)
+   * @param branches The pipelines whose outputs should be combined into a Seq
+   */
+  def gather[A, B : ClassTag](branches: Seq[Pipeline[A, B]]): Pipeline[A, Seq[B]] = {
+    // attach a value per branch to offset all existing node ids by.
+    val branchesWithNodeOffsets = branches.scanLeft(0)(_ + _.nodes.size).zip(branches)
+
+    val newNodes = branches.map(_.nodes).reduceLeft(_ ++ _) :+ new GatherTransformer[B]
+
+    val newDataDeps = branchesWithNodeOffsets.map { case (offset, branch) =>
+      val dataDeps = branch.dataDeps
+      dataDeps.map(_.map(x => if (x == Pipeline.SOURCE) Pipeline.SOURCE else x + offset))
+    }.reduceLeft(_ ++ _) :+  branchesWithNodeOffsets.map { case (offset, branch) =>
+      val sink = branch.sink
+      if (sink == Pipeline.SOURCE) Pipeline.SOURCE else sink + offset
+    }
+
+    val newFitDeps = branchesWithNodeOffsets.map { case (offset, branch) =>
+      val fitDeps = branch.fitDeps
+      fitDeps.map(_.map(x => if (x == Pipeline.SOURCE) Pipeline.SOURCE else x + offset))
+    }.reduceLeft(_ ++ _) :+  Seq()
+
+    val newSink = newNodes.size - 1
+    Pipeline(newNodes, newDataDeps, newFitDeps, newSink)
+  }
+
+}
+
+/**
```

Comment:
Can we pull this out into a new file ? It looks like a big enough class to do that